roms_tools/setup/cdr_release.py

```diff
@@ -0,0 +1,481 @@
+    times : list of datetime.datetime, optional
+        Explicit time points for volume fluxes and tracer concentrations. Defaults to [self.start_time, self.end_time] if None.
+
+        Example: `times=[datetime(2022, 1, 1), datetime(2022, 1, 2), datetime(2022, 1, 3)]`
+
+    tracer_fluxes : dict, optional
+
+        Dictionary of tracer names and their non-negative flux values. The flux values can be either
+        a float (constant in time) or a list of float (time-varying).
+
+        - Constant: applies uniformly across the entire simulation period.
+        - Time-varying: must match the length of `times`.
+
+        Default is an empty dictionary (`{}`) if not provided.
+
+        Example:
+
+        - Constant: `{"ALK": 2000.0, "DIC": 1900.0}`
+        - Time-varying: `{"ALK": [2000.0, 2050.0, 2013.3], "DIC": [1900.0, 1920.0, 1910.2]}`
+        - Mixed: `{"ALK": 2000.0, "DIC": [1900.0, 1920.0, 1910.2]}`
+    """
+
+    times: Optional[List[datetime]] = None
+    tracer_fluxes: Optional[
+        Dict[str, Union[NonNegativeFloat, List[NonNegativeFloat]]]
+    ] = None
+
+    @model_validator(mode="after")
+    def _postprocess(self) -> "TracerPerturbation":
+        if self.times is None:
+            self.times = []
+        num_times = len(self.times)
+
+        if self.tracer_fluxes is None:
+            self.tracer_fluxes = {}
+
+        # Fill all tracer fluxes that are not provided with zero
+        defaults = get_tracer_defaults()
+        for tracer_name in defaults.keys():
+            if tracer_name not in self.tracer_fluxes:
+                self.tracer_fluxes[tracer_name] = 0.0
+
+        self.tracer_fluxes = {
+            tracer: (flux if isinstance(flux, Flux) else Flux(name=tracer, values=flux))
+            for tracer, flux in self.tracer_fluxes.items()
+        }
+
+        for flux in self.tracer_fluxes.values():
+            flux.check_length(num_times)
+
+        return self
+
+    def _extend_to_endpoints(self, start_time, end_time) -> "TracerPerturbation":
+        """Ensures that time series data includes endpoints at `start_time` and
+        `end_time`.
+
+        Pads each tracer flux if needed to match the full time window.
+        Also ensures `self.times` includes the endpoints.
+        """
+        for flux in self.tracer_fluxes.values():
+            flux.extend_to_endpoints(self.times, start_time, end_time)
+        self.extend_times_to_endpoints(start_time, end_time)
+
+    def _simplified_dump(self) -> dict:
+        """Return a simplified dict representation with flattened values."""
+
+        with warnings.catch_warnings():
+            warnings.simplefilter("ignore")
+            data = self.model_dump()
+        data["release_type"] = "TracerPerturbation"
+
+        # Flatten tracer_fluxes
+        if "tracer_fluxes" in data:
+            simplified = {}
+            for tracer, contents in data["tracer_fluxes"].items():
```

Comment:
Can we perform this entire action without a loop using the dataset/xarray? This is potentially garbage pseudo-code:

```
tracer_data = self.ds["tracer_fluxes"].loc[all_tracer_names].flatten() # or some stride thing
```

Reply:
The Release objects don't have access to the xarray Dataset.